tests/MongoDB.Driver.Tests/Specifications/unified-test-format/UnifiedEntityMap.cs

```diff
@@ -0,0 +1,480 @@
+﻿/* Copyright 2020-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using MongoDB.Bson;
+using MongoDB.Driver.Core;
+using MongoDB.Driver.Core.Events;
+using MongoDB.Driver.GridFS;
+using MongoDB.Driver.TestHelpers;
+
+namespace MongoDB.Driver.Tests.Specifications.unified_test_format
+{
+    public sealed class UnifiedEntityMap : IDisposable
+    {
+        // private variables
+        private readonly Dictionary<string, IGridFSBucket> _buckets = new Dictionary<string, IGridFSBucket>();
+        private readonly Dictionary<string, IEnumerator<ChangeStreamDocument<BsonDocument>>> _changeStreams = new Dictionary<string, IEnumerator<ChangeStreamDocument<BsonDocument>>>();
+        private readonly Dictionary<string, EventCapturer> _clientEventCapturers = new Dictionary<string, EventCapturer>();
+        private readonly Dictionary<string, DisposableMongoClient> _clients = new Dictionary<string, DisposableMongoClient>();
+        private readonly Dictionary<string, IMongoCollection<BsonDocument>> _collections = new Dictionary<string, IMongoCollection<BsonDocument>>();
+        private readonly Dictionary<string, IMongoDatabase> _databases = new Dictionary<string, IMongoDatabase>();
+        private readonly Dictionary<string, BsonValue> _results = new Dictionary<string, BsonValue>();
+        private readonly Dictionary<string, IClientSessionHandle> _sessions = new Dictionary<string, IClientSessionHandle>();
+        private readonly Dictionary<string, BsonDocument> _sessionIds = new Dictionary<string, BsonDocument>();
+
+        public UnifiedEntityMap(BsonArray entitiesArray)
```

Comment:
This is a lot of work to do in a constructor.

Consider a static factory method instead:

    public static UnifiedEntityMap Create(BsonArray entitiesArray)

To separate out parsing the entitiesArray from the constructor itself.

Reply:
It can be done with a lot less boilerplate. Here's what I had in mind:

    #region static
    // public static methods
    public static UnifiedEntityMap Create(BsonArray entitiesArray)
    {
        var map = new UnifiedEntityMap();
        if (entitiesArray != null)
        {
            map.Initialize(entitiesArray);
        }
        return map;
    }
    #endregion

    // constructors
    private UnifiedEntityMap()
    {
    }

    // private methods
    private void Initialize(BsonArray entitiesArray)
    {
        // the code that is currently in the public constructor
    }